# Introdução
Este notebook descreve a criação e um classificador que diz se pessoas em imagens estão ou não usando mascara.

As imagens podem retratar zero ou mais pessoas em diferentes lugares e algumas pessoas em uma imagem podem estar utilizando mascaras enquanto outras na mesma imagem não. Por essa característica foi decido criar o classificador a partir da junção de duas redes neurais, uma rede para identificar faces e outra para dizer se essas faces estão utilizando mascaras. Poderia ser utilizado uma única rede neural, de topologia Yolo, para fazer as duas coisas contudo, se presumiu que assim a qualidade das predições seria prejudicada (testes são necessários para confirmar isso).

Para identificar as regiões nas imagens que correspondem a faces foi treinada uma rede neural Yolo 4. Apesar de já existirem versões mais recentes da topologia Yolo as implementações que geram essas topologias usam a licença GPL, que é uma licença open source restritiva e não permitindo portanto a criação de produtos proprietários para códigos que linkam um código *GPL*. Tanto a implementação Yolo utilizada quanto a versão original utilizam licenças open source permissivas, a implementação utilizada neste notebook utiliza a licença *MIT* e a implementação original esta registrada em domínio publico.


Para classificar se a face, em uma imagem região da imagem, esta utilizando mascara foi utilizado uma rede neural com uma topologia próxima da VGG16, possibilitando assim fazer *transfer learning* da rede neural VGG16.








# Download da base de dados

In [ ]:
! rm -rf data
! mkdir data
! export key_to_download_dataset="" ; curl -L "https://universe.roboflow.com/ds/nLPECCY7Z2?key=$key_to_download_dataset" > data/roboflow.zip; cd data; unzip roboflow.zip; rm roboflow.zip

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   902  100   902    0     0   1610      0 --:--:-- --:--:-- --:--:--  1610
100  206M  100  206M    0     0  16.4M      0  0:00:12  0:00:12 --:--:-- 21.8M
Archive:  roboflow.zip
 extracting: README.dataset.txt      
 extracting: README.roboflow.txt     
   creating: test/
 extracting: test/0209-00176-076b1_jpg.rf.68ca9cf32fcefea879d225af4a7c3d44.jpg  
 extracting: test/0450908675_50159485_mutation-virus-chine-inquietude_jpg.rf.bda5fb5e13aac2a13eacab5ee4938218.jpg  
 extracting: test/0_Concern-In-China-As-Mystery-Virus-Spreads_jpg.rf.11f76c39529068ff636fa2242afffa1d.jpg  
 extracting: test/0_Concern-In-China-As-Mystery-Virus-Spreads_jpg.rf.8817cb47eb5195e35bd6eee4bc773e68.jpg  
 extracting: test/1224331650_g_400-w_g_jpg.rf.4f6fa6b1a07d7011f5e32cf441b0d8d8.jpg  
 extracting: test/1224331650_g_400-w_g_jpg.rf.58aaec865ea99dac949f22

In [ ]:
! mkdir models
! mkdir models/callbacks
! mkdir models/callbacks/yolo
! mkdir models/callbacks/yolo/face

In [ ]:
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


# Instação, ajustes no código e criação de funções auxiliares para a Yolo 4

## Download do Yolo 4

In [ ]:
! git clone https://github.com/taipingeric/yolo-v4-tf.keras.git
! wget https://github.com/AlexeyAB/darknet/releases/download/darknet_yolo_v3_optimal/yolov4.weights


Cloning into 'yolo-v4-tf.keras'...
remote: Enumerating objects: 1437, done.
remote: Counting objects: 100% (149/149), done.
remote: Compressing objects: 100% (43/43), done.
remote: Total 1437 (delta 124), reused 128 (delta 106), pack-reused 1288
Receiving objects: 100% (1437/1437), 29.84 MiB | 15.46 MiB/s, done.
Resolving deltas: 100% (750/750), done.
--2023-01-20 04:24:58--  https://github.com/AlexeyAB/darknet/releases/download/darknet_yolo_v3_optimal/yolov4.weights
Resolving github.com (github.com)... 20.205.243.166
Connecting to github.com (github.com)|20.205.243.166|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/75388965/ba4b6380-889c-11ea-9751-f994f5961796?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230120%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230120T042459Z&X-Amz-Expires=300&X-Amz-Signature=265d95468b1c304b7c64bf4251f855fc5552cd374e9d2

## Aplicar mudanças no código da Yolo para retirar algumas mensagens de log, diminuir a taxa de aprendizado e acrescentar data augmentation.

In [ ]:
import os
yolo_patch='''
From aff9bb5606aea8572001e2719ae3a0cac83905c0 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Anderson=20Gon=C3=A7alves=20Marco?=
 <anderson.marco@gmail.com>
Date: Thu, 19 Jan 2023 02:16:45 -0300
Subject: [PATCH 1/4] Small changes

---
 models.py | 6 +++---
 utils.py  | 2 +-
 2 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/models.py b/models.py
index 3ca0cd6..ea33fd2 100755
--- a/models.py
+++ b/models.py
@@ -80,7 +80,7 @@ class Yolov4(object):
                 self.training_model.load_weights(self.weight_path)
                 print(f'load from {self.weight_path}')
 
-        self.training_model.compile(optimizer=optimizers.Adam(lr=1e-3),
+        self.training_model.compile(optimizer=optimizers.Adam(lr=1e-4),
                                     loss={'yolo_loss': lambda y_true, y_pred: y_pred})
 
     def load_model(self, path):
@@ -107,10 +107,10 @@ class Yolov4(object):
                                 initial_epoch=initial_epoch)
     # raw_img: RGB
     def predict_img(self, raw_img, random_color=True, plot_img=True, figsize=(10, 10), show_text=True, return_output=False):
-        print('img shape: ', raw_img.shape)
+        #print('img shape: ', raw_img.shape)
         img = self.preprocess_img(raw_img)
         imgs = np.expand_dims(img, axis=0)
-        pred_output = self.inference_model.predict(imgs)
+        pred_output = self.inference_model.predict(imgs,verbose=False)
         detections = get_detection_data(img=raw_img,
                                         model_outputs=pred_output,
                                         class_names=self.class_names)
diff --git a/utils.py b/utils.py
index 1144290..f31ba5c 100755
--- a/utils.py
+++ b/utils.py
@@ -74,7 +74,7 @@ def get_detection_data(img, model_outputs, class_names):
     df['w'] = df['x2'] - df['x1']
     df['h'] = df['y2'] - df['y1']
 
-    print(f'# of bboxes: {num_bboxes}')
+    #print(f'# of bboxes: {num_bboxes}')
     return df
 
 def read_annotation_lines(annotation_path, test_size=None, random_seed=5566):
-- 
2.25.1


From 8dac3a6e0bed7010ec8f89afb7036f1660ecf520 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Anderson=20Gon=C3=A7alves=20Marco?=
 <anderson.marco@gmail.com>
Date: Thu, 19 Jan 2023 17:45:53 -0300
Subject: [PATCH 2/4] Small changes

---
 utils.py | 26 +++++++++++++++++++++++++-
 1 file changed, 25 insertions(+), 1 deletion(-)

diff --git a/utils.py b/utils.py
index f31ba5c..e741595 100755
--- a/utils.py
+++ b/utils.py
@@ -7,7 +7,8 @@ import os
 from sklearn.model_selection import train_test_split
 from tensorflow.keras.utils import Sequence
 from config import yolo_config
-
+import random
+from skimage.util import random_noise
 
 def load_weights(model, weights_file_path):
     conv_layer_size = 110
@@ -128,6 +129,7 @@ class DataGenerator(Sequence):
                  class_name_path,
                  folder_path,
                  max_boxes=100,
+                 apply_data_generator=False,
                  shuffle=True):
         self.annotation_lines = annotation_lines
         self.class_name_path = class_name_path
@@ -140,6 +142,7 @@ class DataGenerator(Sequence):
         self.indexes = np.arange(len(self.annotation_lines))
         self.folder_path = folder_path
         self.max_boxes = max_boxes
+        self.__apply_data_generator=apply_data_generator
         self.on_epoch_end()
 
     def __len__(self):
@@ -184,10 +187,31 @@ class DataGenerator(Sequence):
 
         return X, y_tensor, y_true_boxes_xywh
 
+    def __data_generator(self,img):
+        if(random.random()<0.1):
+            op_to_choice=random.random()
+            if(op_to_choice<=0.33):
+                print('noise')
+                return random_noise(img)
+            elif(op_to_choice<=0.66):
+                k_size = random.randrange(1,10,2)
+                img_blur = cv2.medianBlur(img, k_size)
+                print('blur')
+                return img_blur
+            else:
+                bright = np.ones(img.shape, dtype="uint8") * (int((random.random()*30))+40)
+                brightincrease = cv2.add(img, bright)
+                print('bright')
+                return brightincrease
+        else:
+            return img
+
     def get_data(self, annotation_line):
         line = annotation_line.split()
         img_path = line[0]
         img = cv2.imread(os.path.join(self.folder_path, img_path))[:, :, ::-1]
+        if(self.__apply_data_generator):
+            img=self.__data_generator(img)
         ih, iw = img.shape[:2]
         h, w, c = self.target_img_size
         boxes = np.array([np.array(list(map(float, box.split(',')))) for box in line[1:]], dtype=np.float32) # x1y1x2y2
-- 
2.25.1


From 043e5b9538335f736fc7462c3d44f7fc2d532e3e Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Anderson=20Gon=C3=A7alves=20Marco?=
 <anderson.marco@gmail.com>
Date: Thu, 19 Jan 2023 17:49:36 -0300
Subject: [PATCH 3/4] Small changes

---
 utils.py | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/utils.py b/utils.py
index e741595..2e2d6c2 100755
--- a/utils.py
+++ b/utils.py
@@ -129,7 +129,7 @@ class DataGenerator(Sequence):
                  class_name_path,
                  folder_path,
                  max_boxes=100,
-                 apply_data_generator=False,
+                 apply_data_augmentation=False,
                  shuffle=True):
         self.annotation_lines = annotation_lines
         self.class_name_path = class_name_path
@@ -142,7 +142,7 @@ class DataGenerator(Sequence):
         self.indexes = np.arange(len(self.annotation_lines))
         self.folder_path = folder_path
         self.max_boxes = max_boxes
-        self.__apply_data_generator=apply_data_generator
+        self.__apply_data_augmentation=apply_data_augmentation
         self.on_epoch_end()
 
     def __len__(self):
@@ -187,7 +187,7 @@ class DataGenerator(Sequence):
 
         return X, y_tensor, y_true_boxes_xywh
 
-    def __data_generator(self,img):
+    def __data_augmentation(self,img):
         if(random.random()<0.1):
             op_to_choice=random.random()
             if(op_to_choice<=0.33):
@@ -210,8 +210,8 @@ class DataGenerator(Sequence):
         line = annotation_line.split()
         img_path = line[0]
         img = cv2.imread(os.path.join(self.folder_path, img_path))[:, :, ::-1]
-        if(self.__apply_data_generator):
-            img=self.__data_generator(img)
+        if(self.__apply_data_augmentation):
+            img=self.__data_augmentation(img)
         ih, iw = img.shape[:2]
         h, w, c = self.target_img_size
         boxes = np.array([np.array(list(map(float, box.split(',')))) for box in line[1:]], dtype=np.float32) # x1y1x2y2
-- 
2.25.1


From df7667d9e820ab4f625f848350f963ec1215d30f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Anderson=20Gon=C3=A7alves=20Marco?=
 <anderson.marco@gmail.com>
Date: Thu, 19 Jan 2023 17:56:52 -0300
Subject: [PATCH 4/4] Small changes

---
 utils.py | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/utils.py b/utils.py
index 2e2d6c2..6637dc9 100755
--- a/utils.py
+++ b/utils.py
@@ -191,17 +191,14 @@ class DataGenerator(Sequence):
         if(random.random()<0.1):
             op_to_choice=random.random()
             if(op_to_choice<=0.33):
-                print('noise')
                 return random_noise(img)
             elif(op_to_choice<=0.66):
                 k_size = random.randrange(1,10,2)
                 img_blur = cv2.medianBlur(img, k_size)
-                print('blur')
                 return img_blur
             else:
                 bright = np.ones(img.shape, dtype="uint8") * (int((random.random()*30))+40)
                 brightincrease = cv2.add(img, bright)
-                print('bright')
                 return brightincrease
         else:
             return img
-- 
2.25.1


'''
with open('patch_yolo.patch','w') as fp:
    fp.write(yolo_patch)
os.system('cd yolo-v4-tf.keras; git apply ../patch_yolo.patch')

0

## Carrega o código que gera modelos Yolo 4

In [ ]:
import sys
if(not('yolo-v4-tf.keras' in  sys.path)):
     sys.path.append("yolo-v4-tf.keras")




## Funcões para auxiliar na geração de arquivos de configuração do datasets enviados para a Yolo 4.

In [ ]:
def create_dict_to_convert_class_name__to_number():
    conv_class_name_to_number={}
    with open('data/train/_classes.txt') as fp:
        class_number=0
        for line in fp:            
            conv_class_name_to_number[line.replace('\n','')]=class_number
            class_number=class_number+1
    return conv_class_name_to_number

def get_annotations_of_a_line_of_annotations_for_a_specific_mask(line,class_number):
    line = line.replace("\n","")
    fields=line.split(" ")
    fields_masks=fields[1:]
    fields_to_return_from_line=[]
    for field in fields_masks:
        sub_fields=field.split(",")
        if(sub_fields[-1]==str(class_number)):
            fields_to_return_from_line.append(field)
    return fields_to_return_from_line
    
def get_image_name_from_a_line_of_annotations_for_a_specific_mask(line):
    line = line.replace("\n","")
    fields=line.split(" ")
    return fields[0]

def filter_images_list_by_class(classes,file_with_img_list='data/train/_annotations.txt'):
    lines_with_annotations_filtered=[]
    with open(file_with_img_list) as fp:        
        for line in fp:
            image_name=get_image_name_from_a_line_of_annotations_for_a_specific_mask(line)
            fields=[]
            for class_number in classes:
                annotations=get_annotations_of_a_line_of_annotations_for_a_specific_mask(line,class_number)
                if(len(annotations)>0):
                    fields=fields+annotations
            if(len(fields)>0):
                fields=[image_name]+fields
                final_line=(" ".join(fields))+"\n"
                lines_with_annotations_filtered.append(final_line)
    return lines_with_annotations_filtered

def replace_class_of_anotation(anotation,new_class):
    fields=anotation.split(",")
    fields[-1]=str(new_class)
    return ",".join(fields)

def change_the_class_list_of_lines_with_annotations(lines_with_annotations,conv_from_old_number_to_new_number):
    classes=list(conv_from_old_number_to_new_number.keys())
    lines_to_return=[]
    for line in lines_with_annotations:
        image_name=get_image_name_from_a_line_of_annotations_for_a_specific_mask(line)
        fields=[]
        for class_number in classes:
            class_to_replace=conv_from_old_number_to_new_number[class_number]
            annotations=get_annotations_of_a_line_of_annotations_for_a_specific_mask(line,class_number)
            if(len(annotations)>0):
                annotations_with_class_replaced=[]
                for anotation in annotations:
                    annotations_with_class_replaced.append(replace_class_of_anotation(anotation,class_to_replace))
                fields=fields+annotations_with_class_replaced
        
        if(len(fields)>0):
            fields=[image_name]+fields
            final_line=(" ".join(fields))+"\n"
            lines_to_return.append(final_line)
    return lines_to_return
                    
def write_list_of_lines_to_a_file(path,lines):
    with open(path,'w') as fp:
        for line in lines:
            fp.write(line)


# Funções necessárias para extrair métricas de qualidade paras as predição do Yolo 4.

In [ ]:
import tqdm
import pandas as pd
class GenerateIds():
    def __init__(self):
        self.__id_generator=0
    
    def next_id(self):
        self.__id_generator= self.__id_generator+1
        return self.__id_generator
    
def point_inside_of_box(point,box):
    if((point['x']>=box['x1'] and point['x']<=box['x2']) and 
       (point['y']>=box['y1'] and point['y']<=box['y2']) 
      ):
        return True
    else:
        return False

def box_overlap(box1,box2):
    points_box1=[{'x':box1['x1'],'y':box1['y1']},
                 {'x':box1['x2'],'y':box1['y1']},
                 {'x':box1['x1'],'y':box1['y2']},
                 {'x':box1['x2'],'y':box1['y2']}]
    
    points_box2=[{'x':box2['x1'],'y':box2['y1']},
                 {'x':box2['x2'],'y':box2['y1']},
                 {'x':box2['x1'],'y':box2['y2']},
                 {'x':box2['x2'],'y':box2['y2']}]
    
    for point in points_box1:
        if(point_inside_of_box(point,box2)):
            return True
        
    for point in points_box2:
        if(point_inside_of_box(point,box1)):
            return True
        
    return False
    
def calc_iou(box1,box2):
    if(box_overlap(box1,box2)==False):
        return -1
    
    xi1=max(box1['x1'],box2['x1'])
    yi1=max(box1['y1'],box2['y1'])
    xi2=min(box1['x2'],box2['x2'])
    yi2=min(box1['y2'],box2['y2'])
    are_box1=(box1['x1']-box1['x2'])*((box1['y1']-box1['y2']))
    are_box2=(box2['x1']-box2['x2'])*((box2['y1']-box2['y2']))
    area_intersection =max((xi2 - xi1)*(yi2 - yi1),0)
    
    union_area = (are_box1 + are_box2) - area_intersection
    return area_intersection/union_area

def generate_pred_boxes_in_dict_from_df_with_pred_boxes(prediction_boxes_in_df,generateIds:GenerateIds):
    list_of_pred_box_in_dict=[]
    uuids_created_in_the_function=[]
    for i in range(len(prediction_boxes_in_df)):
        
        row=prediction_boxes_in_df.iloc[i]
        pred_box={'x1':float(row['x1']),'x2':float(row['x2']),'y1':float(row['y1']),'y2':float(row['y2']),
                  'class_name':row['class_name'],
                  'associated_with_img_box_id':None,
                  'iou_of_img_box_associated':-1,
                  'id': generateIds.next_id()
                 }
        list_of_pred_box_in_dict.append(pred_box)
    return list_of_pred_box_in_dict

def generate_imgs_boxes_in_dict_from_imgs_boxes_in_str_yolo4_format(img_filename,list_of_img_boxes_yolo4_format_str,generateIds:GenerateIds, dict_to_convert_class_id_to_name):
    list_of_img_box_in_dict=[]
    for img_boxes_yolo4_format_str in list_of_img_boxes_yolo4_format_str:
        fields=img_boxes_yolo4_format_str.split(',')
        img_box={'x1':float(fields[0]),'x2':float(fields[2]),
                 'y1':float(fields[1]),'y2':float(fields[3]),
                  'class_name':dict_to_convert_class_id_to_name[fields[4].replace("\n","")],
                  'associated_with_pred_box_id':None,
                  'iou_of_pred_box_associated':-1,
                  'id': generateIds.next_id(),
                  'img_filename':img_filename
                 }
        list_of_img_box_in_dict.append(img_box)
    return list_of_img_box_in_dict
    
def set_better_prediction_box_for_a_image_box(pred_boxes,img_boxes,img_boxes_generated,pred_boxes_generated):
    for pre_box in pred_boxes:
        for img_box in img_boxes:
            iou=calc_iou(pre_box,img_box)
            if(pre_box['iou_of_img_box_associated']<iou and img_box['iou_of_pred_box_associated']<iou):
                
                
                if(pre_box['associated_with_img_box_id']!= None):
                    prev_img_box_asso= img_boxes_generated[pre_box['associated_with_img_box_id']]
                    prev_img_box_asso['associated_with_pred_box_id']=None
                    prev_img_box_asso['iou_of_pred_box_associated']=-1
                
                
                if(img_box['associated_with_pred_box_id']!= None):
                    prev_pred_box_asso= pred_boxes_generated[img_box['associated_with_pred_box_id']]
                    prev_pred_box_asso['associated_with_img_box_id']=None
                    prev_pred_box_asso['iou_of_pred_box_associated']=-1
                
                pre_box['iou_of_img_box_associated']=iou
                img_box['iou_of_pred_box_associated']=iou
                pre_box['associated_with_img_box_id']=img_box['id']
                img_box['associated_with_pred_box_id']=pre_box['id']
            
        
def transform_pred_boxes_and_imgs_box_to_dfs(pred_boxes_generated,img_boxes_generated):
    ids=list(pred_boxes_generated.keys())
    df_pred_box_in_dict={'id_pred_box':[],'iou':[],'class_pred_box':[],'class_img_box':[],'id_img_box':[]}
    df_img_box_in_dict={'img':[],'id_pred_box':[],'iou':[],'class_pred_box':[],'class_img_box':[],'id_img_box':[],
                       'x1_pred':[],'x2_pred':[],'y1_pred':[],'y2_pred':[]}
    for id0 in ids:
        prd_box=pred_boxes_generated[id0]
        id_img_box=prd_box['associated_with_img_box_id']
        if(id_img_box!=None):
            img_box=img_boxes_generated[id_img_box]
            df_pred_box_in_dict['class_img_box'].append(img_box['class_name'])
        else:
            df_pred_box_in_dict['class_img_box'].append('-1')
            
        df_pred_box_in_dict['id_pred_box'].append(prd_box['id'])
        df_pred_box_in_dict['iou'].append(prd_box['iou_of_img_box_associated'])
        df_pred_box_in_dict['class_pred_box'].append(prd_box['class_name'])
        
        df_pred_box_in_dict['id_img_box'].append(prd_box['associated_with_img_box_id'])
    
    ids=list(img_boxes_generated.keys())
    for id0 in ids:        
        img_box=img_boxes_generated[id0]
        id_prd_box=img_box['associated_with_pred_box_id']
        if(id_prd_box!=None):
            prd_box=pred_boxes_generated[id_prd_box]
            df_img_box_in_dict['class_pred_box'].append(prd_box['class_name'])
            df_img_box_in_dict['x1_pred'].append(prd_box['x1'])
            df_img_box_in_dict['x2_pred'].append(prd_box['x2'])
            df_img_box_in_dict['y1_pred'].append(prd_box['y1'])
            df_img_box_in_dict['y2_pred'].append(prd_box['y2'])
        else:
            df_img_box_in_dict['class_pred_box'].append('-1')
            df_img_box_in_dict['x1_pred'].append(-1)
            df_img_box_in_dict['x2_pred'].append(-1)
            df_img_box_in_dict['y1_pred'].append(-1)
            df_img_box_in_dict['y2_pred'].append(-1)
            
        df_img_box_in_dict['id_pred_box'].append(prd_box['id'])
        df_img_box_in_dict['iou'].append(prd_box['iou_of_img_box_associated'])
        
        df_img_box_in_dict['class_img_box'].append(img_box['class_name'])
        df_img_box_in_dict['id_img_box'].append(img_box['id'])
        df_img_box_in_dict['img'].append(img_box['img_filename'])
    
    df_pred_boxes=pd.DataFrame.from_dict(df_pred_box_in_dict)
    df_img_boxes=pd.DataFrame.from_dict(df_img_box_in_dict)
    
    return {'df_pred_boxes':df_pred_boxes,'df_img_boxes':df_img_boxes }

def apply_metric(model,lines_with_imgs_path_and_annot_in_yolo_v4_format,
                 path_to_dir_where_are_stored_the_images_pointed_by_lines,
                 conv_class_number_to_class_name={'0':'face'}):
    path_to_dir=path_to_dir_where_are_stored_the_images_pointed_by_lines
    img_boxes_generated={}
    pred_boxes_generated={}
    generate_id=GenerateIds()
    for img_data_in_str in tqdm.tqdm(lines_with_imgs_path_and_annot_in_yolo_v4_format):        
        img_data_in_str=img_data_in_str.replace("\n","")
        img_fields=img_data_in_str.split(" ")
        list_of_img_boxes_yolo4_format_str=img_fields[1:]
        list_img_boxes=generate_imgs_boxes_in_dict_from_imgs_boxes_in_str_yolo4_format(img_fields[0],list_of_img_boxes_yolo4_format_str,generate_id,conv_class_number_to_class_name)
        for img_box in list_img_boxes:
            img_boxes_generated[img_box['id']]=img_box
            
        predicted_boxes=model.predict(path_to_dir+img_fields[0],plot_img=False,show_text=False)
        list_pred_boxes=generate_pred_boxes_in_dict_from_df_with_pred_boxes(predicted_boxes,generate_id)
        for img_box in list_pred_boxes:
            pred_boxes_generated[img_box['id']]=img_box
            
        set_better_prediction_box_for_a_image_box(list_pred_boxes,list_img_boxes,img_boxes_generated,pred_boxes_generated)
        #print(img_boxes_generated)
   
    return transform_pred_boxes_and_imgs_box_to_dfs(pred_boxes_generated,img_boxes_generated)


# Criar um modelo Yolo 4 para capturar a face das imagens

Cria copias das anotações dos datasets de treino e validação, nestas copias as *labels* **mascara** e **não mascara** vão ser a *label* **face**

In [ ]:
conv=create_dict_to_convert_class_name__to_number()
conv_to_new_class={0:0,1:0}
conv=create_dict_to_convert_class_name__to_number()
lines_with_annotations_filtered=filter_images_list_by_class([0,1],file_with_img_list='data/train/_annotations.txt')
lines_with_annotations_filtered2=change_the_class_list_of_lines_with_annotations(lines_with_annotations_filtered,conv_to_new_class)
write_list_of_lines_to_a_file('train_annotation.txt',lines_with_annotations_filtered2)

lines_with_annotations_filtered=filter_images_list_by_class([0,1],file_with_img_list='data/valid/_annotations.txt')
lines_with_annotations_filtered2=change_the_class_list_of_lines_with_annotations(lines_with_annotations_filtered,conv_to_new_class)
write_list_of_lines_to_a_file('valid_annotation.txt',lines_with_annotations_filtered2)


FOLDER_PATH = 'data/train'
class_name_path = 'classes.txt'
with open('train_annotation.txt') as fp:
    train_lines= fp.readlines()
with open('valid_annotation.txt') as fp:
    val_lines= fp.readlines()


Cria o arquivo que diz quais classes podem ter os boxes inferidas pelo Yolo

In [ ]:
!echo "face" > classes.txt 


## Criar modelo sem data augmentation

Cria um modelo Yolo 4 para a detecção de face na imagem.

In [ ]:
from utils import DataGenerator, read_annotation_lines
from models import Yolov4
import tensorflow as tf
FOLDER_PATH = 'data/train'
class_name_path = 'classes.txt'
data_gen_train = DataGenerator(train_lines, 
                               class_name_path,                              
                               'data/train',
                               apply_data_augmentation=False)
data_gen_val = DataGenerator(val_lines, 
                             class_name_path, 
                             'data/valid')
model = Yolov4(weight_path='yolov4.weights', 
               class_name_path=class_name_path)

model_checkpoint_callback = tf.keras.callbacks.ModelCheckpoint(
    filepath='models/yolo_face_without_data_augmentation_callback.h5',
    save_weights_only=False,
    monitor='val_loss',
    mode='min',
    save_best_only=True)

model.fit(data_gen_train, 
          initial_epoch=0,
          epochs=20, 
          val_data_gen=data_gen_val,
          callbacks=[model_checkpoint_callback])
model.save_model('models/yolo_face_without_data_augmentation.h5')

nms iou: 0.413 score: 0.3
failed to read  all weights, # of unread weights: 0
load from yolov4.weights
Epoch 1/20
147/147 [==============================] - 121s 713ms/step - loss: 18712.6973 - val_loss: nan
Epoch 2/20
147/147 [==============================] - 101s 690ms/step - loss: 7852.6846 - val_loss: nan
Epoch 3/20
147/147 [==============================] - 103s 703ms/step - loss: 3060.9192 - val_loss: nan
Epoch 4/20
147/147 [==============================] - 107s 725ms/step - loss: 1396.9431 - val_loss: 968.2948
Epoch 5/20
147/147 [==============================] - 108s 731ms/step - loss: 749.3418 - val_loss: 601.2958
Epoch 6/20
147/147 [==============================] - 107s 729ms/step - loss: 452.1957 - val_loss: 416.7884
Epoch 7/20
147/147 [==============================] - 107s 728ms/step - loss: 295.3052 - val_loss: 315.5072
Epoch 8/20
147/147 [==============================] - 107s 725ms/step - loss: 207.7248 - val_loss: 254.6887
Epoch 9/20
147/147 [=======================

In [ ]:
! #cp models/yolo_face_without_data_augmentation.h5 drive/MyDrive/Colab\ Notebooks/visao_computacional/yolo_face_without_data_augmentation.h5
! #cp models/yolo_face_without_data_augmentation_callback.h5 drive/MyDrive/Colab\ Notebooks/visao_computacional/yolo_face_without_data_augmentation_callback.h5

## Criar modelo com data augmentation

In [ ]:
from utils import DataGenerator, read_annotation_lines
from models import Yolov4
import tensorflow as tf
FOLDER_PATH = 'data/train'
class_name_path = 'classes.txt'
data_gen_train = DataGenerator(train_lines, 
                               class_name_path,                              
                               'data/train',
                               apply_data_augmentation=True)
data_gen_val = DataGenerator(val_lines, 
                             class_name_path, 
                             'data/valid')
model = Yolov4(weight_path='yolov4.weights', 
               class_name_path=class_name_path)

model_checkpoint_callback = tf.keras.callbacks.ModelCheckpoint(
    filepath='models/yolo_face_with_data_augmentation_callback.h5',
    save_weights_only=False,
    monitor='val_loss',
    mode='min',
    save_best_only=True)

model.fit(data_gen_train, 
          initial_epoch=0,
          epochs=20, 
          val_data_gen=data_gen_val,
          callbacks=[model_checkpoint_callback])
model.save_model('models/yolo_face_with_data_augmentation.h5')

nms iou: 0.413 score: 0.3
failed to read  all weights, # of unread weights: 0
load from yolov4.weights


/usr/local/lib/python3.8/dist-packages/keras/optimizers/optimizer_v2/adam.py:110: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.
  super(Adam, self).__init__(name, **kwargs)


Epoch 1/20
147/147 [==============================] - 139s 753ms/step - loss: 18708.5703 - val_loss: nan
Epoch 2/20
147/147 [==============================] - 113s 766ms/step - loss: 7862.9985 - val_loss: nan
Epoch 3/20
147/147 [==============================] - 109s 728ms/step - loss: 3059.0513 - val_loss: nan
Epoch 4/20
147/147 [==============================] - 111s 753ms/step - loss: 1408.8229 - val_loss: 1023.5786
Epoch 5/20
147/147 [==============================] - 117s 792ms/step - loss: 776.8868 - val_loss: 614.3392
Epoch 6/20
147/147 [==============================] - 114s 773ms/step - loss: 502.7813 - val_loss: 415.6751
Epoch 7/20
147/147 [==============================] - 114s 774ms/step - loss: 318.6432 - val_loss: 295.0506
Epoch 8/20
147/147 [==============================] - 111s 756ms/step - loss: 226.5023 - val_loss: 244.0314
Epoch 9/20
147/147 [==============================] - 124s 844ms/step - loss: 181.8727 - val_loss: 202.3535
Epoch 10/20
147/147 [================

In [ ]:
! cp models/yolo_face_with_data_augmentation.h5 drive/MyDrive/Colab\ Notebooks/visao_computacional/yolo_face_with_data_augmentation.h5
! cp models/yolo_face_with_data_augmentation_callback.h5 drive/MyDrive/Colab\ Notebooks/visao_computacional/yolo_face_with_data_augmentation_callback.h5

## Comparar métricas da Yolo 4 gerada por data augmentation e sem  data augmentation

### Yolo 4 sem data augmentation

In [ ]:
! cp drive/MyDrive/Colab\ Notebooks/visao_computacional/yolo_face_without_data_augmentation_callback.h5  models/yolo_face_without_data_augmentation_callback.h5 

In [ ]:
from models import Yolov4
model2 = Yolov4( class_name_path=class_name_path)
model2.yolo_model.load_weights('models/yolo_face_without_data_augmentation_callback.h5')

nms iou: 0.413 score: 0.3


Código para extrair as métricas de qualidade das imagens

Aplicar as métricas na base de validação

In [ ]:
dfs_with_metrics=apply_metric(model2,val_lines,'data/valid/')

100%|██████████| 312/312 [00:43<00:00,  7.11it/s]


Caixas de imagens que não foram preditas pela Yolo 4

In [ ]:
df1=dfs_with_metrics['df_img_boxes']
len(df1[df1['class_pred_box']=='-1'])

136

Caixas de imagens que foram preditas pela Yolo 4

In [ ]:
df1=dfs_with_metrics['df_pred_boxes']
len(df1[df1['class_img_box']!='-1'])

573

Caixas de imagens preditas pela Yolo 4 que não existem na base de dados, chamadas aqui de *caixas sem correspondencia* 

In [ ]:
df1=dfs_with_metrics['df_pred_boxes']
len(df1[df1['class_img_box']=='-1'])

100

Média e desvio padrão de quão próximas caixas preditas estão próximas,em termos de formato, das reais:

In [ ]:
df1=dfs_with_metrics['df_pred_boxes']
df1=df1[df1['class_img_box']!='-1']
print('média: '+str(df1['iou'].mean()))
print('desvio padrão: '+str(df1['iou'].std()))

média: 0.6762442009311177
desvio padrão: 0.19358878726615839


### Yolo 4 com data augmentation

In [ ]:
! cp drive/MyDrive/Colab\ Notebooks/visao_computacional/yolo_face_with_data_augmentation_callback.h5  models/yolo_face_with_data_augmentation_callback.h5 

In [ ]:
from models import Yolov4
model2 = Yolov4( class_name_path=class_name_path)
model2.yolo_model.load_weights('models/yolo_face_with_data_augmentation_callback.h5')

nms iou: 0.413 score: 0.3


/usr/local/lib/python3.8/dist-packages/keras/optimizers/optimizer_v2/adam.py:110: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.
  super(Adam, self).__init__(name, **kwargs)


Código para extrair as métricas de qualidade das imagens

Aplicar as métricas na base de validação




In [ ]:
dfs_with_metrics=apply_metric(model2,val_lines,'data/valid/')

100%|██████████| 312/312 [00:43<00:00,  7.15it/s]


Caixas de imagens que não foram preditas pela Yolo 4

In [ ]:
df1=dfs_with_metrics['df_img_boxes']
len(df1[df1['class_pred_box']=='-1'])

114

Caixas de imagens que foram preditas pela Yolo 4

In [ ]:
df1=dfs_with_metrics['df_pred_boxes']
len(df1[df1['class_img_box']!='-1'])

595

Caixas de imagens preditas pela Yolo 4 que não existem na base de dados,
chamadas aqui de **caixas sem correspondência* * 


In [ ]:
df1=dfs_with_metrics['df_pred_boxes']
len(df1[df1['class_img_box']=='-1'])

177

Média e desvio padrão de quão próximas caixas preditas estão próximas, em termos de formato, das reais:

In [ ]:
df1=dfs_with_metrics['df_pred_boxes']
df1=df1[df1['class_img_box']!='-1']
print('média: '+str(df1['iou'].mean()))
print('desvio padrão: '+str(df1['iou'].std()))

média: 0.6402993600432209
desvio padrão: 0.19544357851147617


### Conlusão sobre a qualidade dos modelos Yolo 4 com e sem data augmentation

O modelo Yolo 4 com *data augmentation* foi o melhor na identificação de caixas de imagens que existem no conjunto de validação contudo, ele gerou mais *caixas sem correspondência*  que o modelo treinado sem *data augmentation*.  Portanto o modelo Yolo 4 treinado sem *data augmentation* apresentou melhor qualidade e ele vai o ser utilizado em conjunto pelo classificador de faces.

# Criar um modelo para classificar faces em com ou sem mascara

Criar um dataframe  com as imagens de treino e as regiões onde estão as faces nas imagens, neste dataframe vai ser utilizado as caixas de imagens fornecidas pelo dataset

In [ ]:
import pandas as pd
def generate_df_imgs_of_faces_from_annotation_lines(lines):
    df_in_dict={'img':[],'x1':[],'x2':[],'y1':[],'y2':[],'class':[]}
    for line in lines:
        line=line.replace("\n","")
        fields=line.split(" ")
        annotations=fields[1:]
        img=fields[0]
        for annotation in annotations:
            elements_annotation=annotation.split(",")
            df_in_dict['img'].append(img)
            df_in_dict['x1'].append(int(elements_annotation[0]))
            df_in_dict['y1'].append(int(elements_annotation[1]))
            df_in_dict['x2'].append(int(elements_annotation[2]))
            df_in_dict['y2'].append(int(elements_annotation[3]))
            df_in_dict['class'].append(int(elements_annotation[4]))
    
    return pd.DataFrame.from_dict(df_in_dict)
    
with open('data/train/_annotations.txt') as fp:
    train_lines=fp.readlines()



df_imgs_train=generate_df_imgs_of_faces_from_annotation_lines(train_lines)


Modelo Yolo utilizado para inferir as regiões onde estão os rostos nas imagens.

In [ ]:
! cp drive/MyDrive/Colab\ Notebooks/visao_computacional/yolo_face_without_data_augmentation_callback.h5  models/yolo_to_use_in_classification_faces.h5 

In [ ]:
!echo "face" > classes.txt 


In [ ]:
from models import Yolov4
class_name_path='classes.txt'
model2 = Yolov4( class_name_path=class_name_path)
model2.yolo_model.load_weights('models/yolo_to_use_in_classification_faces.h5')

nms iou: 0.413 score: 0.3


/usr/local/lib/python3.8/dist-packages/keras/optimizers/optimizer_v2/adam.py:110: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.
  super(Adam, self).__init__(name, **kwargs)


Criar um dataframe  com as imagens de validação e as regiões onde estão as faces nas imagens, neste dataframe vai ser utilizado as caixas de imagens inferidas pelo modelo Yolo

In [ ]:
def create_df_to_use_in_data_generator_from_df_of_metrics_computed_for_yolo_model(df_with_metrics):
    df_with_metrics['x1']=df_with_metrics['x1_pred'].apply(lambda x: int(x))
    df_with_metrics['x2']=df_with_metrics['x2_pred'].apply(lambda x: int(x))
    df_with_metrics['y1']=df_with_metrics['y1_pred'].apply(lambda x: int(x))
    df_with_metrics['y2']=df_with_metrics['y2_pred'].apply(lambda x: int(x))
    df_with_metrics['class']=df_with_metrics['class_img_box']
    df_with_metrics=df_with_metrics[df_with_metrics['x1']!=-1]
    return df_with_metrics


with open('data/valid/_annotations.txt') as fp:
    valid_lines=fp.readlines()
    
dfs_with_metrics=apply_metric(model2,valid_lines,'data/valid/',conv_class_number_to_class_name={'0':0,'1':1})
df_validate_img_boxes= dfs_with_metrics['df_img_boxes']
df_imgs_valid=create_df_to_use_in_data_generator_from_df_of_metrics_computed_for_yolo_model(df_validate_img_boxes)

100%|██████████| 368/368 [01:01<00:00,  5.98it/s]


In [ ]:
import cv2
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Dense
from tensorflow.keras.layers import Dropout
from tensorflow.keras.layers import BatchNormalization
from tensorflow.keras.layers import Conv1D
from tensorflow.keras.layers import LSTM
from tensorflow.keras.layers import GRU
from tensorflow.keras.layers import Flatten
from tensorflow.keras.layers import GlobalMaxPooling1D
from tensorflow.keras.layers import MaxPool1D
from tensorflow.keras.layers import Embedding, Input,Concatenate
from tensorflow.keras.utils import Sequence
from tensorflow.keras.preprocessing.text import one_hot
from tensorflow.keras.preprocessing.sequence import pad_sequences
import numpy as np
import random
import tqdm
import cv2
from skimage import io, img_as_ubyte
from skimage.transform import rotate, AffineTransform, warp
from skimage.util import random_noise


class DataGenerator(Sequence):
    'Generates data for Keras'
    def __init__(   self,
                    df_train,
                    image_dim,
                    pre_process_function_name,
                    dir_where_images_are_stored,
                    apply_data_augmentation=False,
                    batch_size=16,
                    shuffle=True):
        
        self.__batch_size = batch_size
        
        self.shuffle = shuffle
        self.__random_state=0
        self.__image_dim=image_dim
        self.__df_train=df_train
        self.__dir_where_images_are_stored=dir_where_images_are_stored
        self.__pre_process_funcition_name=pre_process_function_name
        self.__apply_data_augmentation=apply_data_augmentation
        self.on_epoch_end()

    
    def get_image_dim(self):
        return self.__image_dim
    

    def get_batch_size(self):
        return self.__batch_size


    def __len__(self):
        'Denotes the number of batches per epoch'
        return int(np.floor(len(self.__df_train) / self.__batch_size))

    def __data_augmentation(self,img):
        if(random.random()<0.2):
            op_to_choice=random.random()
            if(op_to_choice<=0.125):
                return random_noise(img)
            elif(op_to_choice<=0.25):
                k_size = random.randrange(1,10,2)
                img_blur = cv2.medianBlur(img, k_size)
                return img_blur
            elif(op_to_choice<=0.375):
                angle = random.randint(0, 180)
                return rotate(img, -angle)
            elif(op_to_choice<=0.50):
                angle = random.randint(0, 180)
                return rotate(img, angle)
            elif(op_to_choice<=0.625):
                return np.fliplr(img)
            elif(op_to_choice<=0.75):
                return np.flipud(img)
            elif(op_to_choice<=0.875):
                try:
                    zoom_value = random.random()
                    hidth, width = img.shape[:2]
                    h_taken = int(zoom_value*hidth)
                    w_taken = int(zoom_value*width)
                    h_start = random.randint(0, hidth-h_taken)
                    w_start = random.randint(0, width-w_taken)
                    img2 = img[h_start:h_start+h_taken, w_start:w_start+w_taken, :]
                    img2 = cv2.resize(img2, (hidth, width), cv2.INTER_CUBIC)
                    return img2
                except:
                   return img
            else:
                bright = np.ones(img.shape, dtype="uint8") * (int((random.random()*30))+40)
                brightincrease = cv2.add(img, bright)
                return brightincrease
        else:
            return img
        
    def __load_image_as_matrix(self,pre_process_funcition_name,image_filename,x1,y1,x2,y2):
        im1=cv2.imread(image_filename)
        im1=im1[y1:y2,x1:x2]
        if(self.__apply_data_augmentation):
            im1=self.__data_augmentation(im1)
        im1=cv2.resize(im1,(self.__image_dim[1],self.__image_dim[0]),interpolation=cv2.INTER_AREA)
        if(pre_process_funcition_name==''):
            im1=((np.array(im1,dtype=float)/255)-0.5)*2
            return im1
        elif(
            pre_process_funcition_name=='VGG16' or 
            pre_process_funcition_name=='VGG19' or
            pre_process_funcition_name=='keras.applications.vgg16' or
            pre_process_funcition_name=='inception_v3'):
            
            im2=im1.copy()
            im2[:,:,0]=im1[:,:,2]
            im2[:,:,2]=im1[:,:,0]
            im2=np.array(im2,dtype=np.float32)
            im2=np.expand_dims(im2, axis=0)
            if(
                pre_process_funcition_name=='VGG16'):
                    im3=tf.keras.applications.vgg16.preprocess_input(im2)[0,:,:,:]
                    return im3
            elif(pre_process_funcition_name=="inception_v3"):
                im3=tf.keras.applications.inception_v3.preprocess_input(im2)[0,:,:,:]
                return im3
            elif(pre_process_funcition_name=="VGG19"):
                im3=tf.keras.applications.vgg19.preprocess_input(im2)[0,:,:,:]
                return im3
            else:
                raise Exception("Invalid pre function name") 

        else:
            raise Exception("Invalid pre function name") 

                
    
  

    def __getitem__(self, index):
        'Generate one batch of data'
        # Generate indexes of the batch
        
        df=self.__df_train
        df_batch=df[index*self.__batch_size:(index+1)*self.__batch_size]
        X=np.zeros((len(df_batch),self.__image_dim[0],self.__image_dim[1],3),dtype=float)
        y=np.zeros((len(df_batch),2),dtype=float)
        for i in range(len(df_batch)):
            row=df_batch.iloc[i]
            imgpath=self.__dir_where_images_are_stored+'/'+row['img']
            imgpath=imgpath.replace("\n","")
            img= self.__load_image_as_matrix(self.__pre_process_funcition_name,imgpath,row['x1'],row['y1'],row['x2'],row['y2'])
            label= df_batch.iloc[i]['class']
            X[i,:,:,:]=img[:,:,:]
            y[i,label]=1.0


        
        return X, y

    def on_epoch_end(self):
        'Updates indexes after each epoch'
        
        if self.shuffle == True:
            self.__df_train=self.__df_train.sample(n=len(self.__df_train), random_state=self.__random_state)

        self.__random_state=self.__random_state+1

In [ ]:
def create_model(img_dim):
   
    base_model = VGG16(include_top=False,weights='imagenet', input_shape=(img_dim[0],img_dim[1],3))
    flat1 = Flatten()(base_model.layers[-1].output)
    drop1=Dropout(0.2)(flat1)
    dense1= Dense(1024, activation='relu')(drop1)
    drop2=Dropout(0.2)(dense1)
    dense2 = Dense(2, activation='softmax')(drop2)
    model = Model(inputs=base_model.inputs, outputs=dense2)
    opt=tf.keras.optimizers.Adam(learning_rate=0.0001)
    model.compile(loss='categorical_crossentropy', optimizer=opt,metrics='accuracy')
    return model

#Treina o modelo utilizando *data_augmentation* 

In [ ]:
from tensorflow.keras.applications.vgg16 import VGG16
import tensorflow.keras.applications.vgg16
from tensorflow.keras.applications.vgg19 import VGG19
import tensorflow.keras.applications.vgg19
from tensorflow.keras.applications.inception_v3 import InceptionV3
import tensorflow.keras.applications.inception_v3
from tensorflow.keras.applications import InceptionResNetV2

preprocess_input="VGG16"
img_dim=(224,224)
datagenerator_train=DataGenerator(df_imgs_train,img_dim,preprocess_input,'data/train',apply_data_augmentation=True)
model=create_model(img_dim)
datagenerator_valid=DataGenerator(df_imgs_valid,img_dim,preprocess_input,'data/valid')


58889256/58889256 [==============================] - 0s 0us/step


In [ ]:
model_checkpoint_callback = tf.keras.callbacks.ModelCheckpoint(
    filepath='models/classification_with_data_augmentation_callback.h5',
    save_weights_only=False,
    monitor='val_accuracy',
    mode='max',
    save_best_only=True)

model.fit(datagenerator_train,
        validation_data=datagenerator_valid,
        epochs=20,use_multiprocessing=True,
        workers=8,
        callbacks=[model_checkpoint_callback])

Epoch 1/20
161/161 [==============================] - 82s 460ms/step - loss: 0.5935 - accuracy: 0.7620 - val_loss: 0.2104 - val_accuracy: 0.9143
Epoch 2/20
161/161 [==============================] - 74s 441ms/step - loss: 0.2907 - accuracy: 0.8657 - val_loss: 0.2272 - val_accuracy: 0.9125
Epoch 3/20
161/161 [==============================] - 84s 501ms/step - loss: 0.2218 - accuracy: 0.9107 - val_loss: 0.2139 - val_accuracy: 0.9214
Epoch 4/20
161/161 [==============================] - 71s 428ms/step - loss: 0.1996 - accuracy: 0.9231 - val_loss: 0.1891 - val_accuracy: 0.9107
Epoch 5/20
161/161 [==============================] - 81s 483ms/step - loss: 0.1601 - accuracy: 0.9286 - val_loss: 0.1766 - val_accuracy: 0.9304
Epoch 6/20
161/161 [==============================] - 78s 465ms/step - loss: 0.1358 - accuracy: 0.9340 - val_loss: 0.3203 - val_accuracy: 0.9107
Epoch 7/20
161/161 [==============================] - 80s 480ms/step - loss: 0.1472 - accuracy: 0.9402 - val_loss: 0.2946 - val_ac

In [ ]:
!cp models/classification_with_data_augmentation_callback.h5 drive/MyDrive/Colab\ Notebooks/visao_computacional/classification_with_data_augmentation_callback.h5

#Treina o modelo sem *data_augmentation* 

In [ ]:
from tensorflow.keras.applications.vgg16 import VGG16
import tensorflow.keras.applications.vgg16
from tensorflow.keras.applications.vgg19 import VGG19
import tensorflow.keras.applications.vgg19
from tensorflow.keras.applications.inception_v3 import InceptionV3
import tensorflow.keras.applications.inception_v3
from tensorflow.keras.applications import InceptionResNetV2

preprocess_input="VGG16"
img_dim=(224,224)
datagenerator_train=DataGenerator(df_imgs_train,img_dim,preprocess_input,'data/train',apply_data_augmentation=False)
model=create_model(img_dim)
datagenerator_valid=DataGenerator(df_imgs_valid,img_dim,preprocess_input,'data/valid')


58889256/58889256 [==============================] - 3s 0us/step


In [ ]:
model_checkpoint_callback = tf.keras.callbacks.ModelCheckpoint(
    filepath='models/classification_without_data_augmentation_callback.h5',
    save_weights_only=False,
    monitor='val_accuracy',
    mode='max',
    save_best_only=True)

model.fit(datagenerator_train,
        validation_data=datagenerator_valid,
        epochs=20,use_multiprocessing=True,
        workers=8,
        callbacks=[model_checkpoint_callback])

Epoch 1/20
161/161 [==============================] - 77s 422ms/step - loss: 0.5562 - accuracy: 0.7904 - val_loss: 0.2864 - val_accuracy: 0.8964
Epoch 2/20
161/161 [==============================] - 74s 442ms/step - loss: 0.1939 - accuracy: 0.9356 - val_loss: 0.3574 - val_accuracy: 0.8714
Epoch 3/20
161/161 [==============================] - 71s 422ms/step - loss: 0.1098 - accuracy: 0.9651 - val_loss: 0.2153 - val_accuracy: 0.9161
Epoch 4/20
161/161 [==============================] - 70s 419ms/step - loss: 0.1150 - accuracy: 0.9596 - val_loss: 0.7231 - val_accuracy: 0.8089
Epoch 5/20
161/161 [==============================] - 72s 429ms/step - loss: 0.0996 - accuracy: 0.9620 - val_loss: 0.2624 - val_accuracy: 0.9089
Epoch 6/20
161/161 [==============================] - 68s 402ms/step - loss: 0.0691 - accuracy: 0.9775 - val_loss: 0.2585 - val_accuracy: 0.9036
Epoch 7/20
161/161 [==============================] - 67s 405ms/step - loss: 0.0406 - accuracy: 0.9876 - val_loss: 0.2866 - val_ac

In [ ]:
! cp models/classification_without_data_augmentation_callback.h5 drive/MyDrive/Colab\ Notebooks/visao_computacional/classification_without_data_augmentation_callback.h5

## Análise de qualidade das redes neurais de classificação no conjunto de validação
Código para realizar predições

In [ ]:
def create_dataframe_with_class_predictions(data_generator: DataGenerator,model):
    df_with_pred__dict_format={'y':[],'y_pred':[],'prob_to_be_one':[]}
    for i in tqdm.tqdm(range(data_generator.__len__())):
        x,y=data_generator.__getitem__(i)
        y1=model.predict(x,verbose=False)
        for j in range(len(y1)):
            if(y[j,1]==1.0):
                df_with_pred__dict_format['y'].append(1.0)
            else:
                df_with_pred__dict_format['y'].append(0.0)
            if(y1[j,1]>0.5):
                df_with_pred__dict_format['y_pred'].append(1.0)
            else:
                df_with_pred__dict_format['y_pred'].append(0.0)
            df_with_pred__dict_format['prob_to_be_one'].append(y1[j,1])
    return pd.DataFrame.from_dict(df_with_pred__dict_format)



### Analisar modelo gerado por *data_augmentation*
Carregar o modelo




In [ ]:
!cp drive/MyDrive/Colab\ Notebooks/visao_computacional/classification_with_data_augmentation_callback.h5 models/classification_with_data_augmentation_callback.h5

In [ ]:
from tensorflow.keras.applications.vgg16 import VGG16
import tensorflow.keras.applications.vgg16
from tensorflow.keras.applications.vgg19 import VGG19
import tensorflow.keras.applications.vgg19
from tensorflow.keras.applications.inception_v3 import InceptionV3
import tensorflow.keras.applications.inception_v3
from tensorflow.keras.applications import InceptionResNetV2

preprocess_input="VGG16"
img_dim=(224,224)
model=create_model(img_dim)
model.load_weights('models/classification_with_data_augmentation_callback.h5')

Realiza as predições no conjunto de validação


In [ ]:
df_prediction=create_dataframe_with_class_predictions(datagenerator_valid,model)

100%|██████████| 35/35 [00:13<00:00,  2.62it/s]


Área sob a curva ROC

In [ ]:
from sklearn.metrics import roc_auc_score 
roc_auc_score(df_prediction['y'], df_prediction['prob_to_be_one'])

0.9868244000897062

Matriz de confusão



In [ ]:
from sklearn.metrics import confusion_matrix
pd.DataFrame(confusion_matrix(df_prediction['y'],df_prediction['y_pred']),columns=[['predito','predito'],['Com mascara','Sem mascara']], index=[['real','real'],['Com mascara','Sem mascara']])

predito            
                 Com mascara Sem mascara
real Com mascara         338          26
     Sem mascara           9         187

### Analisar modelo gerado sem *data augmentation*
Carregar o modelo


In [ ]:
!cp drive/MyDrive/Colab\ Notebooks/visao_computacional/classification_without_data_augmentation_callback.h5 models/classification_without_data_augmentation_callback.h5

In [ ]:
from tensorflow.keras.applications.vgg16 import VGG16
import tensorflow.keras.applications.vgg16
from tensorflow.keras.applications.vgg19 import VGG19
import tensorflow.keras.applications.vgg19
from tensorflow.keras.applications.inception_v3 import InceptionV3
import tensorflow.keras.applications.inception_v3
from tensorflow.keras.applications import InceptionResNetV2

preprocess_input="VGG16"
img_dim=(224,224)
model=create_model(img_dim)
model.load_weights('models/classification_without_data_augmentation_callback.h5')

Realiza as predições no conjunto de validação


In [ ]:
df_prediction=create_dataframe_with_class_predictions(datagenerator_valid,model)

100%|██████████| 35/35 [00:14<00:00,  2.41it/s]


Área sob a curva ROC

In [ ]:
from sklearn.metrics import roc_auc_score 
roc_auc_score(df_prediction['y'], df_prediction['prob_to_be_one'])

0.9791292890782686

Matriz de confusão



In [ ]:
from sklearn.metrics import confusion_matrix
pd.DataFrame(confusion_matrix(df_prediction['y'],df_prediction['y_pred']),columns=[['predito','predito'],['Com mascara','Sem mascara']], index=[['real','real'],['Com mascara','Sem mascara']])

predito            
                 Com mascara Sem mascara
real Com mascara         343          21
     Sem mascara          10         186

## Conclusão sobre as métricas obtidas pelos modelos de classificação

Ambos os modelos apresentam resultados parecidos. Para relizar a análise de qualidade no conjunto de testes vai ser utilizado o modelo com *data augmentation* porque ele apresenta um AUC (área sob a curva ROC) melhor.

# Análise da qualidade das duas redes neurais (Yolo e classificação) no conjunto de testes

## Carregar modelos Yolo e de classificação

In [ ]:
! cp drive/MyDrive/Colab\ Notebooks/visao_computacional/classification_with_data_augmentation_callback.h5 models/classification_model_to_use.h5
! cp drive/MyDrive/Colab\ Notebooks/visao_computacional/yolo_face_without_data_augmentation_callback.h5  models/yolo_to_use_in_classification_faces.h5 
! echo "face" > classes.txt 

In [ ]:
from models import Yolov4
class_name_path='classes.txt'
model_yolo = Yolov4( class_name_path=class_name_path)
model_yolo.yolo_model.load_weights('models/yolo_to_use_in_classification_faces.h5')

from tensorflow.keras.applications.vgg16 import VGG16
import tensorflow.keras.applications.vgg16
from tensorflow.keras.applications.vgg19 import VGG19
import tensorflow.keras.applications.vgg19
from tensorflow.keras.applications.inception_v3 import InceptionV3
import tensorflow.keras.applications.inception_v3
from tensorflow.keras.applications import InceptionResNetV2

preprocess_input="VGG16"
img_dim=(224,224)
model=create_model(img_dim)
model.load_weights('models/classification_model_to_use.h5')

nms iou: 0.413 score: 0.3


/usr/local/lib/python3.8/dist-packages/keras/optimizers/optimizer_v2/adam.py:110: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.
  super(Adam, self).__init__(name, **kwargs)


## Métricas de qualidade para Yolo 4

Aplicar os testes na base de testes

In [ ]:
conv=create_dict_to_convert_class_name__to_number()
conv_to_new_class={0:0,1:0}
conv=create_dict_to_convert_class_name__to_number()

lines_with_annotations_filtered=filter_images_list_by_class([0,1],file_with_img_list='data/test/_annotations.txt')
lines_with_annotations_filtered2=change_the_class_list_of_lines_with_annotations(lines_with_annotations_filtered,conv_to_new_class)
write_list_of_lines_to_a_file('test_annotation.txt',lines_with_annotations_filtered2)
with open('test_annotation.txt') as fp:
    test_lines=fp.readlines()
    
dfs_with_metrics=apply_metric(model_yolo,test_lines,'data/test/')

100%|██████████| 165/165 [00:23<00:00,  7.10it/s]


Caixas de imagens que não foram preditas pela Yolo 4

In [ ]:
df1=dfs_with_metrics['df_img_boxes']
len(df1[df1['class_pred_box']=='-1'])

52

Caixas de imagens que foram preditas pela Yolo 4

In [ ]:
df1=dfs_with_metrics['df_pred_boxes']
len(df1[df1['class_img_box']!='-1'])

325

Caixas de imagens preditas pela Yolo 4 que não existem na base de dados, chamadas aqui de *caixas sem correspondencia* 

In [ ]:
df1=dfs_with_metrics['df_pred_boxes']
len(df1[df1['class_img_box']=='-1'])

65

Média e desvio padrão de quão próximas caixas preditas estão próximas, em termos de formato, das reais

In [ ]:
df1=dfs_with_metrics['df_pred_boxes']
df1=df1[df1['class_img_box']!='-1']
print('média: '+str(df1['iou'].mean()))
print('desvio padrão: '+str(df1['iou'].std()))

média: 0.6137406290692635
desvio padrão: 0.21166768815790704


## Métricas de qualidade para modelo de classificação

In [ ]:
def create_df_to_use_in_data_generator_from_df_of_metrics_computed_for_yolo_model(df_with_metrics):
    df_with_metrics['x1']=df_with_metrics['x1_pred'].apply(lambda x: int(x))
    df_with_metrics['x2']=df_with_metrics['x2_pred'].apply(lambda x: int(x))
    df_with_metrics['y1']=df_with_metrics['y1_pred'].apply(lambda x: int(x))
    df_with_metrics['y2']=df_with_metrics['y2_pred'].apply(lambda x: int(x))
    df_with_metrics['class']=df_with_metrics['class_img_box']
    df_with_metrics=df_with_metrics[df_with_metrics['x1']!=-1]
    return df_with_metrics



with open('data/test/_annotations.txt') as fp:
    test_lines=fp.readlines()
    
dfs_with_metrics=apply_metric(model_yolo,valid_lines,'data/test/',conv_class_number_to_class_name={'0':0,'1':1})
df_test_img_boxes= dfs_with_metrics['df_img_boxes']
df_test=create_df_to_use_in_data_generator_from_df_of_metrics_computed_for_yolo_model(df_test_img_boxes)


class_name_path = 'classes.txt'

datagenerator_test=DataGenerator(df_test,img_dim,preprocess_input,'data/test')
df_prediction=create_dataframe_with_class_predictions(datagenerator_test,model)

100%|██████████| 20/20 [00:07<00:00,  2.59it/s]


Área sob a curva ROC

In [ ]:
from sklearn.metrics import roc_auc_score 
roc_auc_score(df_prediction['y'], df_prediction['prob_to_be_one'])

0.9831002331002331

Matriz de confusão

In [ ]:
from sklearn.metrics import confusion_matrix
pd.DataFrame(confusion_matrix(df_prediction['y'],df_prediction['y_pred']),columns=[['predito','predito'],['Com mascara','Sem mascara']], index=[['real','real'],['Com mascara','Sem mascara']])

predito            
                 Com mascara Sem mascara
real Com mascara         214          28
     Sem mascara           5          73

# Conclusão
Este notebook propos o uso de redes neurais para identificar se pessoas estão utilizando ou não mascaras a partir de imagens. O classificador final é composto por um modelo de rede neural com topologia Yolo, para detectar regiões das imagens onde estão as faces das pessoas, em conjunto de uma rede neural com topologia próxima da VGG16, para detectar se as imagens das faces estão ou não usando mascara.

Com relação as tecnicas de treinamento foram testados treinamento com e sem *data augmentation*. Para a rede neural Yolo o melhor resultado foi alcançado sem *data augmentation*. Para a rede neural de classificação de imagens o melhor resultado foi alcançado com *data augmentation*.

Com relação as métricas de qualidade no conunto de testes fica observado que:
* O modelo Yolo conseguiu identificar 86% das faces.
* A rede neural que determinada se face esta ou não usando mascara apresenta uma acurácia próxima de 90% e um AUC de 0.98. 


Pode-se concluir então que o classificador, criado a partir das duas redes neurais, apresenta uma boa qualidade.



